lib/src/metta/text.rs

```diff
@@ -77,19 +77,27 @@ impl<'a> SExprParser<'a> {
                 '$' => {
                     self.it.next();
                     let token = next_var(&mut self.it);
-                    return Some(Atom::var(token));
+                    return Ok(Some(Atom::var(token)));
                 },
                 '(' => {
                     self.it.next();
-                    return Some(self.parse_expr(tokenizer))
+                    if let Ok(expr) = self.parse_expr(tokenizer) {
+                        return Ok(Some(expr));
+                    } else {
+                        return Err("Error parsing expression".to_string());
+                    }
                 },
-                ')' => panic!("Unexpected right bracket"),
+                ')' => return Err("Unexpected right bracket".to_string()),
                 _ => {
-                    return self.parse_atom(tokenizer);
+                    if let Some(atom) = self.parse_atom(tokenizer) {
+                        return Ok(Some(atom));
+                    } else {
+                        return Err("Error parsing atom".to_string());
+                    }
```

Comment:
Can be simplified to:
```suggestion
                    return Ok(self.parse_atom(tokenizer));
```